In [32]:
#import libraies
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [24]:
Ball_by_ball= pd.read_csv('Ball_By_Ball.csv')
match=pd.read_csv('Match.csv')
Player=pd.read_csv('Player_excel.csv')
Player_match=pd.read_csv('Player_match_excel1.csv',encoding= 'unicode_escape') 

Team=pd.read_csv('Team.csv')                          

In [207]:
##imported file in excel
df=pd.read_excel('batting_avg.xlsx')
df1=pd.read_excel('strike_rate.xlsx')
df2=pd.read_excel('captains.xlsx')
df3=pd.read_excel('fours_and_sixes.xlsx')

In [218]:
df3

name       country  Batter  runs_from_fours  runs from sixes
0        W Jaffer         India       7               56               18
1      MV Boucher  South Africa      11              128               78
2      MEK Hussey     Australia      19              792              312
3        JDP Oram   New Zealand      22               24               30
4       SR Watson     Australia      32             1028              732
..            ...           ...     ...              ...              ...
283    TG Southee   New Zealand     377               12               18
284       D Wiese  South Africa     401               48               24
285  Ishan Kishan         India     410              136               78
286        P Sahu         India     412                4                6
287      K Rabada  South Africa     482               20                6

[288 rows x 5 columns]

In [212]:
df1.columns=['player_name','player_country','strike_rate','Batter','runs']

In [217]:
df3.columns=['name','country','Batter','runs_from_fours','runs from sixes']

In [5]:
Ball_by_ball.head()

MatcH_id  Over_id  Ball_id  Innings_No  Team_Batting  Team_Bowling  \
0    598028       15        6           1             5             2   
1    598028       14        1           1             5             2   
2    598028       14        2           1             5             2   
3    598028       14        3           1             5             2   
4    598028       14        4           1             5             2   

   Striker_Batting_Position Extra_Type  Runs_Scored  Extra_runs  ...  \
0                       6.0  No Extras            4           0  ...   
1                       5.0  No Extras            1           0  ...   
2                       3.0  No Extras            1           0  ...   
3                       5.0  No Extras            1           0  ...   
4                       3.0  No Extras            0           0  ...   

   Fielder_match_SK  Fielder_SK  Bowler_match_SK  BOWLER_SK  \
0                -1          -1            20343         82   
1                -1          -1            20348        345   
2                -1          -1            20348        345   
3                -1          -1            20348        345   
4                -1          -1            20348        345   

   PlayerOut_match_SK  BattingTeam_SK BowlingTeam_SK  Keeper_Catch  \
0                  -1               4              1             0   
1                  -1               4              1             0   
2                  -1               4              1             0   
3                  -1               4              1             0   
4                  -1               4              1             0   

   Player_out_sk  MatchDateSK  
0              0     20130420  
1              0     20130420  
2              0     20130420  
3              0     20130420  
4              0     20130420  

[5 rows x 48 columns]

In [8]:
match.head()

Match_SK  match_id                        Team1  \
0         0    335987  Royal Challengers Bangalore   
1         1    335988              Kings XI Punjab   
2         2    335989             Delhi Daredevils   
3         3    335990               Mumbai Indians   
4         4    335991        Kolkata Knight Riders   

                         Team2 match_date  Season_Year  \
0        Kolkata Knight Riders  4/18/2008         2008   
1          Chennai Super Kings  4/19/2008         2008   
2             Rajasthan Royals  4/19/2008         2008   
3  Royal Challengers Bangalore  4/20/2008         2008   
4              Deccan Chargers  4/20/2008         2008   

                                   Venue_Name   City_Name Country_Name  \
0                       M Chinnaswamy Stadium   Bangalore        India   
1  Punjab Cricket Association Stadium, Mohali  Chandigarh        India   
2                            Feroz Shah Kotla       Delhi        India   
3                            Wankhede Stadium      Mumbai        India   
4                                Eden Gardens     Kolkata        India   

                   Toss_Winner                 match_winner Toss_Name  \
0  Royal Challengers Bangalore        Kolkata Knight Riders     field   
1          Chennai Super Kings          Chennai Super Kings       bat   
2             Rajasthan Royals             Delhi Daredevils       bat   
3               Mumbai Indians  Royal Challengers Bangalore       bat   
4              Deccan Chargers        Kolkata Knight Riders       bat   

  Win_Type Outcome_Type    ManOfMach  Win_Margin  Country_id  
0     runs       Result  BB McCullum       140.0           1  
1     runs       Result   MEK Hussey        33.0           1  
2  wickets       Result  MF Maharoof         9.0           1  
3  wickets       Result   MV Boucher         5.0           1  
4  wickets       Result    DJ Hussey         5.0           1

In [12]:
team.head()

Team_SK  Team_Id                    Team_Name
0        0        1        Kolkata Knight Riders
1        1        2  Royal Challengers Bangalore
2        2        3          Chennai Super Kings
3        3        4              Kings XI Punjab
4        4        5             Rajasthan Royals

In [6]:
Ball_by_ball.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150451 entries, 0 to 150450
Data columns (total 48 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   MatcH_id                  150451 non-null  int64  
 1   Over_id                   150451 non-null  int64  
 2   Ball_id                   150451 non-null  int64  
 3   Innings_No                150451 non-null  int64  
 4   Team_Batting              150451 non-null  int64  
 5   Team_Bowling              150451 non-null  int64  
 6   Striker_Batting_Position  136590 non-null  float64
 7   Extra_Type                150451 non-null  object 
 8   Runs_Scored               150451 non-null  int64  
 9   Extra_runs                150451 non-null  int64  
 10  Wides                     150451 non-null  int64  
 11  Legbyes                   150451 non-null  int64  
 12  Byes                      150451 non-null  int64  
 13  Noballs                   150451 non-null  i

In [14]:
Player.head()

PLAYER_SK  Player_Id      Player_Name         DOB    Batting_hand  \
0          0          1       SC Ganguly  07-08-1972   Left-hand bat   
1          1          2      BB McCullum   9/27/1981  Right-hand bat   
2          2          3       RT Ponting  12/19/1974  Right-hand bat   
3          3          4        DJ Hussey   7/15/1977  Right-hand bat   
4          4          5  Mohammad Hafeez  10/17/1980  Right-hand bat   

        Bowling_skill Country_Name  
0    Right-arm medium        India  
1    Right-arm medium  New Zealand  
2    Right-arm medium    Australia  
3  Right-arm offbreak    Australia  
4  Right-arm offbreak     Pakistan

In [ ]:
Player_match.head()

## DATA CLEANING OF BALL_BY_BALL DATASET

#summary/conclusion
1 total null count
.Player_Out                  143013
 Fielders                    145100 
 Striker_Batting_Position     13861
2.RangeIndex: 150451
3. duplicate=0
4.total match=636
5


In [25]:
Ball_by_ball.head()

MatcH_id  Over_id  Ball_id  Innings_No  Team_Batting  Team_Bowling  \
0    598028       15        6           1             5             2   
1    598028       14        1           1             5             2   
2    598028       14        2           1             5             2   
3    598028       14        3           1             5             2   
4    598028       14        4           1             5             2   

   Striker_Batting_Position Extra_Type  Runs_Scored  Extra_runs  ...  \
0                       6.0  No Extras            4           0  ...   
1                       5.0  No Extras            1           0  ...   
2                       3.0  No Extras            1           0  ...   
3                       5.0  No Extras            1           0  ...   
4                       3.0  No Extras            0           0  ...   

   Fielder_match_SK  Fielder_SK  Bowler_match_SK  BOWLER_SK  \
0                -1          -1            20343         82   
1                -1          -1            20348        345   
2                -1          -1            20348        345   
3                -1          -1            20348        345   
4                -1          -1            20348        345   

   PlayerOut_match_SK  BattingTeam_SK BowlingTeam_SK  Keeper_Catch  \
0                  -1               4              1             0   
1                  -1               4              1             0   
2                  -1               4              1             0   
3                  -1               4              1             0   
4                  -1               4              1             0   

   Player_out_sk  MatchDateSK  
0              0     20130420  
1              0     20130420  
2              0     20130420  
3              0     20130420  
4              0     20130420  

[5 rows x 48 columns]

In [26]:
Ball_by_ball.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150451 entries, 0 to 150450
Data columns (total 48 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   MatcH_id                  150451 non-null  int64  
 1   Over_id                   150451 non-null  int64  
 2   Ball_id                   150451 non-null  int64  
 3   Innings_No                150451 non-null  int64  
 4   Team_Batting              150451 non-null  int64  
 5   Team_Bowling              150451 non-null  int64  
 6   Striker_Batting_Position  136590 non-null  float64
 7   Extra_Type                150451 non-null  object 
 8   Runs_Scored               150451 non-null  int64  
 9   Extra_runs                150451 non-null  int64  
 10  Wides                     150451 non-null  int64  
 11  Legbyes                   150451 non-null  int64  
 12  Byes                      150451 non-null  int64  
 13  Noballs                   150451 non-null  i

In [27]:
Ball_by_ball.describe()

MatcH_id        Over_id        Ball_id     Innings_No  \
count  1.504510e+05  150451.000000  150451.000000  150451.000000   
mean   6.362075e+05      10.142704       3.616639       1.482190   
std    2.343623e+05       5.674255       1.807638       0.501768   
min    3.359870e+05       1.000000       1.000000       1.000000   
25%    4.191540e+05       5.000000       2.000000       1.000000   
50%    5.483820e+05      10.000000       4.000000       1.000000   
75%    8.297420e+05      15.000000       5.000000       2.000000   
max    1.082650e+06      20.000000       9.000000       4.000000   

        Team_Batting   Team_Bowling  Striker_Batting_Position    Runs_Scored  \
count  150451.000000  150451.000000             136590.000000  150451.000000   
mean        5.346272       5.332806                  3.583637       1.222199   
std         3.217641       3.219607                  2.145090       1.594311   
min         1.000000       1.000000                  1.000000       0.000000   
25%         3.000000       3.000000                  2.000000       0.000000   
50%         5.000000       5.000000                  3.000000       1.000000   
75%         7.000000       7.000000                  5.000000       1.000000   
max        13.000000      13.000000                 11.000000       6.000000   

          Extra_runs          Wides  ...  Fielder_match_SK     Fielder_SK  \
count  150451.000000  150451.000000  ...     150451.000000  150451.000000   
mean        0.068993       0.037501  ...        689.975108       4.526730   
std         0.349256       0.257405  ...       3698.806643      37.215835   
min         0.000000       0.000000  ...         -1.000000      -1.000000   
25%         0.000000       0.000000  ...         -1.000000      -1.000000   
50%         0.000000       0.000000  ...         -1.000000      -1.000000   
75%         0.000000       0.000000  ...         -1.000000      -1.000000   
max         5.000000       5.000000  ...      26680.000000     496.000000   

       Bowler_match_SK      BOWLER_SK  PlayerOut_match_SK  BattingTeam_SK  \
count    150451.000000  150451.000000       150451.000000   150451.000000   
mean      19676.764555     193.087098          970.324278        4.346272   
std        4026.168474     132.998950         4353.992290        3.217641   
min       12697.000000       0.000000           -1.000000        0.000000   
25%       16175.000000      76.000000           -1.000000        2.000000   
50%       19674.000000     173.000000           -1.000000        4.000000   
75%       23131.000000     309.000000           -1.000000        6.000000   
max       26685.000000     496.000000        26685.000000       12.000000   

       BowlingTeam_SK   Keeper_Catch  Player_out_sk   MatchDateSK  
count   150451.000000  150451.000000  150451.000000  1.504510e+05  
mean         4.332806       0.000432       1.100704  2.012529e+07  
std          3.219607       0.020781      20.404263  2.763159e+04  
min          0.000000       0.000000      -1.000000  2.008042e+07  
25%          2.000000       0.000000       0.000000  2.010041e+07  
50%          4.000000       0.000000       0.000000  2.012052e+07  
75%          6.000000       0.000000       0.000000  2.015042e+07  
max         12.000000       1.000000     496.000000  2.017052e+07  

[8 rows x 45 columns]

In [28]:
Ball_by_ball.isnull().sum()

MatcH_id                         0
Over_id                          0
Ball_id                          0
Innings_No                       0
Team_Batting                     0
Team_Bowling                     0
Striker_Batting_Position     13861
Extra_Type                       0
Runs_Scored                      0
Extra_runs                       0
Wides                            0
Legbyes                          0
Byes                             0
Noballs                          0
Penalty                          0
Bowler_Extras                    0
Out_type                         0
Caught                           0
Bowled                           0
Run_out                          0
LBW                              0
Retired_hurt                     0
Stumped                          0
caught_and_bowled                0
hit_wicket                       0
ObstructingFeild                 0
Bowler_Wicket                    0
Match_Date                       0
Season              

In [31]:
Ball_by_ball[Ball_by_ball.duplicated()]

Empty DataFrame
Columns: [MatcH_id, Over_id, Ball_id, Innings_No, Team_Batting, Team_Bowling, Striker_Batting_Position, Extra_Type, Runs_Scored, Extra_runs, Wides, Legbyes, Byes, Noballs, Penalty, Bowler_Extras, Out_type, Caught, Bowled, Run_out, LBW, Retired_hurt, Stumped, caught_and_bowled, hit_wicket, ObstructingFeild, Bowler_Wicket, Match_Date, Season, Striker, Non_Striker, Bowler, Player_Out, Fielders, Striker_match_SK, StrikerSK, NonStriker_match_SK, NONStriker_SK, Fielder_match_SK, Fielder_SK, Bowler_match_SK, BOWLER_SK, PlayerOut_match_SK, BattingTeam_SK, BowlingTeam_SK, Keeper_Catch, Player_out_sk, MatchDateSK]
Index: []

[0 rows x 48 columns]

In [33]:
Ball_by_ball['Over_id'].unique()

array([15, 14, 13, 12, 11, 10,  3,  2,  1, 16, 17, 18,  9,  8,  7, 20,  4,
        5,  6, 19], dtype=int64)

In [38]:
Ball_by_ball['Ball_id'].unique()

array([6, 1, 2, 3, 4, 5, 7, 8, 9], dtype=int64)

In [39]:
Ball_by_ball['Innings_No'].unique()

array([1, 2, 4, 3], dtype=int64)

In [49]:
Ball_by_ball['MatcH_id'].nunique()



636

In [ ]:
Ball_by_ball['MatcH_id'].nunique()

In [73]:
Ball_by_ball['Non_Striker'].nunique()

457

In [58]:
sum(Ball_by_ball['Caught']==1)

4373

In [61]:
sum(Ball_by_ball['Stumped']==1)

243

In [85]:
Ball_by_ball['Extra_Type'].unique()

array(['No Extras', 'wides', 'legbyes', 'byes', 'noballs', 'Legbyes',
       'Byes', 'Wides', 'penalty', 'Noballs'], dtype=object)

##list of metric ##
Number of Innings Played as a Batsman (Innings)
Number of Runs Scored (Runs)
Number of Outs/Dismissals (Dismissals)
Number of Balls Faced (Balls Faced)
Number of 4s scored (4s)
Number of 6s scored (6s)
Highest Score (High Score)
Number of Half Centuries scored (50)
Number of Centuries scored (100)
Batting Strike Rate (Strike Rate)
Batting Average (Batting Average)

In [125]:

dfBall = Ball_by_ball.groupby('Striker')\
            .agg({'MatcH_id':'nunique','Runs_Scored':'sum','Ball_id':'count','Wides':'sum','Legbyes':'sum','Byes':'sum','Noballs':'sum'})\
            .reset_index()

In [126]:
dfBall

Striker  MatcH_id  Runs_Scored  Ball_id  Wides  Legbyes  Byes  Noballs
0          1        56         1349     1326     77       30     2        6
1          2       103         2755     2181    128       55     9       18
2          3         9           91      129      2        7     4        0
3          4        61         1322     1101     33       15     9        5
4          5         8           64       84      1        3     1        0
..       ...       ...          ...      ...    ...      ...   ...      ...
455      490         1            3        2      0        0     0        0
456      491         3           38       31      0        0     0        0
457      495         3            8       11      0        1     0        0
458      496         3           19       26      0        0     0        0
459      497         5            9       12      0        1     2        0

[460 rows x 8 columns]

In [127]:
dfBall.columns = ['Batter','Innings','Total_Runs','total_ball','wides','legbyes','byes','noball']

In [128]:
dfBall

Batter  Innings  Total_Runs  total_ball  wides  legbyes  byes  noball
0         1       56        1349        1326     77       30     2       6
1         2      103        2755        2181    128       55     9      18
2         3        9          91         129      2        7     4       0
3         4       61        1322        1101     33       15     9       5
4         5        8          64          84      1        3     1       0
..      ...      ...         ...         ...    ...      ...   ...     ...
455     490        1           3           2      0        0     0       0
456     491        3          38          31      0        0     0       0
457     495        3           8          11      0        1     0       0
458     496        3          19          26      0        0     0       0
459     497        5           9          12      0        1     2       0

[460 rows x 8 columns]

## merging and adding new column in for batter 

In [181]:
dfBall['ball_faced']=dfBall['total_ball']-dfBall['wides']-dfBall['legbyes']-dfBall['byes']-dfBall['noball']

In [182]:
dfBall

Batter  Innings  Total_Runs  total_ball  wides  legbyes  byes  noball  \
0         1       56        1349        1326     77       30     2       6   
1         2      103        2755        2181    128       55     9      18   
2         3        9          91         129      2        7     4       0   
3         4       61        1322        1101     33       15     9       5   
4         5        8          64          84      1        3     1       0   
..      ...      ...         ...         ...    ...      ...   ...     ...   
455     490        1           3           2      0        0     0       0   
456     491        3          38          31      0        0     0       0   
457     495        3           8          11      0        1     0       0   
458     496        3          19          26      0        0     0       0   
459     497        5           9          12      0        1     2       0   

     ball_faced  
0          1211  
1          1971  
2           116  
3          1039  
4            79  
..          ...  
455           2  
456          31  
457          10  
458          26  
459           9  

[460 rows x 9 columns]

In [184]:
dfBall.drop(['wides','legbyes','byes','noball','total_ball'],axis=1)

Batter  Innings  Total_Runs  ball_faced
0         1       56        1349        1211
1         2      103        2755        1971
2         3        9          91         116
3         4       61        1322        1039
4         5        8          64          79
..      ...      ...         ...         ...
455     490        1           3           2
456     491        3          38          31
457     495        3           8          10
458     496        3          19          26
459     497        5           9           9

[460 rows x 4 columns]

In [187]:
dfBall['batting_strike_rate']=((dfBall['Total_Runs']/dfBall['ball_faced'])*100).round(decimals=2)

In [188]:
dfBall

Batter  Innings  Total_Runs  total_ball  wides  legbyes  byes  noball  \
0         1       56        1349        1326     77       30     2       6   
1         2      103        2755        2181    128       55     9      18   
2         3        9          91         129      2        7     4       0   
3         4       61        1322        1101     33       15     9       5   
4         5        8          64          84      1        3     1       0   
..      ...      ...         ...         ...    ...      ...   ...     ...   
455     490        1           3           2      0        0     0       0   
456     491        3          38          31      0        0     0       0   
457     495        3           8          11      0        1     0       0   
458     496        3          19          26      0        0     0       0   
459     497        5           9          12      0        1     2       0   

     ball_faced  batting_strike_rate  
0          1211               111.40  
1          1971               139.78  
2           116                78.45  
3          1039               127.24  
4            79                81.01  
..          ...                  ...  
455           2               150.00  
456          31               122.58  
457          10                80.00  
458          26                73.08  
459           9               100.00  

[460 rows x 10 columns]

In [189]:
## total number of type player out
dfout = Ball_by_ball[(Ball_by_ball['Out_type']!="Not Applicable" )].groupby('Striker')['MatcH_id'].count().reset_index()

In [190]:
dfout

Striker  MatcH_id
0          1        54
1          2       100
2          3         8
3          4        49
4          5         8
..       ...       ...
425      488        12
426      489         2
427      491         2
428      495         3
429      496         3

[430 rows x 2 columns]

In [193]:
dfout.columns=['Batter','out_inning']

In [194]:
dfout

Batter  out_inning
0         1          54
1         2         100
2         3           8
3         4          49
4         5           8
..      ...         ...
425     488          12
426     489           2
427     491           2
428     495           3
429     496           3

[430 rows x 2 columns]

In [196]:
dfBall2=dfBall.merge(dfout,on='Batter',how='left')

In [197]:
dfBall2

Batter  Innings  Total_Runs  total_ball  wides  legbyes  byes  noball  \
0         1       56        1349        1326     77       30     2       6   
1         2      103        2755        2181    128       55     9      18   
2         3        9          91         129      2        7     4       0   
3         4       61        1322        1101     33       15     9       5   
4         5        8          64          84      1        3     1       0   
..      ...      ...         ...         ...    ...      ...   ...     ...   
455     490        1           3           2      0        0     0       0   
456     491        3          38          31      0        0     0       0   
457     495        3           8          11      0        1     0       0   
458     496        3          19          26      0        0     0       0   
459     497        5           9          12      0        1     2       0   

     ball_faced  batting_strike_rate  out_inning  
0          1211               111.40        54.0  
1          1971               139.78       100.0  
2           116                78.45         8.0  
3          1039               127.24        49.0  
4            79                81.01         8.0  
..          ...                  ...         ...  
455           2               150.00         NaN  
456          31               122.58         2.0  
457          10                80.00         3.0  
458          26                73.08         3.0  
459           9               100.00         NaN  

[460 rows x 11 columns]

In [198]:
dfBall2['batting_avg']=(dfBall2['Total_Runs']/dfBall2['out_inning']).round(decimals=2)

In [199]:
dfBall2

Batter  Innings  Total_Runs  total_ball  wides  legbyes  byes  noball  \
0         1       56        1349        1326     77       30     2       6   
1         2      103        2755        2181    128       55     9      18   
2         3        9          91         129      2        7     4       0   
3         4       61        1322        1101     33       15     9       5   
4         5        8          64          84      1        3     1       0   
..      ...      ...         ...         ...    ...      ...   ...     ...   
455     490        1           3           2      0        0     0       0   
456     491        3          38          31      0        0     0       0   
457     495        3           8          11      0        1     0       0   
458     496        3          19          26      0        0     0       0   
459     497        5           9          12      0        1     2       0   

     ball_faced  batting_strike_rate  out_inning  batting_avg  
0          1211               111.40        54.0        24.98  
1          1971               139.78       100.0        27.55  
2           116                78.45         8.0        11.38  
3          1039               127.24        49.0        26.98  
4            79                81.01         8.0         8.00  
..          ...                  ...         ...          ...  
455           2               150.00         NaN          NaN  
456          31               122.58         2.0        19.00  
457          10                80.00         3.0         2.67  
458          26                73.08         3.0         6.33  
459           9               100.00         NaN          NaN  

[460 rows x 12 columns]

In [200]:
dfBall2.drop(['wides','legbyes','byes','noball','total_ball'],axis=1,inplace=True)

In [201]:
dfBall2

Batter  Innings  Total_Runs  ball_faced  batting_strike_rate  out_inning  \
0         1       56        1349        1211               111.40        54.0   
1         2      103        2755        1971               139.78       100.0   
2         3        9          91         116                78.45         8.0   
3         4       61        1322        1039               127.24        49.0   
4         5        8          64          79                81.01         8.0   
..      ...      ...         ...         ...                  ...         ...   
455     490        1           3           2               150.00         NaN   
456     491        3          38          31               122.58         2.0   
457     495        3           8          10                80.00         3.0   
458     496        3          19          26                73.08         3.0   
459     497        5           9           9               100.00         NaN   

     batting_avg  
0          24.98  
1          27.55  
2          11.38  
3          26.98  
4           8.00  
..           ...  
455          NaN  
456        19.00  
457         2.67  
458         6.33  
459          NaN  

[460 rows x 7 columns]

In [214]:
## dfball3 is merge of dfball2 and strike dataset
dfBall3=dfBall2.merge(df1[['Batter','player_name','player_country']],on='Batter',how='left')

In [215]:
dfBall3

Batter  Innings  Total_Runs  ball_faced  batting_strike_rate  out_inning  \
0         1       56        1349        1211               111.40        54.0   
1         2      103        2755        1971               139.78       100.0   
2         3        9          91         116                78.45         8.0   
3         4       61        1322        1039               127.24        49.0   
4         5        8          64          79                81.01         8.0   
..      ...      ...         ...         ...                  ...         ...   
455     490        1           3           2               150.00         NaN   
456     491        3          38          31               122.58         2.0   
457     495        3           8          10                80.00         3.0   
458     496        3          19          26                73.08         3.0   
459     497        5           9           9               100.00         NaN   

     batting_avg        player_name player_country  
0          24.98         SC Ganguly          India  
1          27.55        BB McCullum    New Zealand  
2          11.38         RT Ponting      Australia  
3          26.98          DJ Hussey      Australia  
4           8.00    Mohammad Hafeez       Pakistan  
..           ...                ...            ...  
455          NaN          RD Chahar          India  
456        19.00         SP Jackson          India  
457         2.67           TS Mills        England  
458         6.33       Vishnu Vinod          India  
459          NaN  Washington Sundar          India  

[460 rows x 9 columns]

In [220]:
dfBall4=dfBall3.merge(df3[['Batter','runs_from_fours','runs from sixes']],on='Batter',how='left')

In [221]:
dfBall4

Batter  Innings  Total_Runs  ball_faced  batting_strike_rate  out_inning  \
0         1       56        1349        1211               111.40        54.0   
1         2      103        2755        1971               139.78       100.0   
2         3        9          91         116                78.45         8.0   
3         4       61        1322        1039               127.24        49.0   
4         5        8          64          79                81.01         8.0   
..      ...      ...         ...         ...                  ...         ...   
455     490        1           3           2               150.00         NaN   
456     491        3          38          31               122.58         2.0   
457     495        3           8          10                80.00         3.0   
458     496        3          19          26                73.08         3.0   
459     497        5           9           9               100.00         NaN   

     batting_avg        player_name player_country  runs_from_fours  \
0          24.98         SC Ganguly          India            548.0   
1          27.55        BB McCullum    New Zealand           1108.0   
2          11.38         RT Ponting      Australia             20.0   
3          26.98          DJ Hussey      Australia            360.0   
4           8.00    Mohammad Hafeez       Pakistan             28.0   
..           ...                ...            ...              ...   
455          NaN          RD Chahar          India              NaN   
456        19.00         SP Jackson          India              NaN   
457         2.67           TS Mills        England              NaN   
458         6.33       Vishnu Vinod          India              4.0   
459          NaN  Washington Sundar          India              NaN   

     runs from sixes  
0              252.0  
1              744.0  
2               12.0  
3              360.0  
4               12.0  
..               ...  
455              NaN  
456              NaN  
457              NaN  
458              6.0  
459              NaN  

[460 rows x 11 columns]

In [222]:
dfBall4['boundry_run%']=(((dfBall4['runs_from_fours']+dfBall4["runs from sixes"])/dfBall4['Total_Runs'])*100).round(decimals=2)

In [223]:
dfBall4

Batter  Innings  Total_Runs  ball_faced  batting_strike_rate  out_inning  \
0         1       56        1349        1211               111.40        54.0   
1         2      103        2755        1971               139.78       100.0   
2         3        9          91         116                78.45         8.0   
3         4       61        1322        1039               127.24        49.0   
4         5        8          64          79                81.01         8.0   
..      ...      ...         ...         ...                  ...         ...   
455     490        1           3           2               150.00         NaN   
456     491        3          38          31               122.58         2.0   
457     495        3           8          10                80.00         3.0   
458     496        3          19          26                73.08         3.0   
459     497        5           9           9               100.00         NaN   

     batting_avg        player_name player_country  runs_from_fours  \
0          24.98         SC Ganguly          India            548.0   
1          27.55        BB McCullum    New Zealand           1108.0   
2          11.38         RT Ponting      Australia             20.0   
3          26.98          DJ Hussey      Australia            360.0   
4           8.00    Mohammad Hafeez       Pakistan             28.0   
..           ...                ...            ...              ...   
455          NaN          RD Chahar          India              NaN   
456        19.00         SP Jackson          India              NaN   
457         2.67           TS Mills        England              NaN   
458         6.33       Vishnu Vinod          India              4.0   
459          NaN  Washington Sundar          India              NaN   

     runs from sixes  boundry_run%  
0              252.0         59.30  
1              744.0         67.22  
2               12.0         35.16  
3              360.0         54.46  
4               12.0         62.50  
..               ...           ...  
455              NaN           NaN  
456              NaN           NaN  
457              NaN           NaN  
458              6.0         52.63  
459              NaN           NaN  

[460 rows x 12 columns]

In [224]:
dfBall4.to_csv('player_batting_dataset')

## We use the same aggregation methods with a combination of conditional slicing by excluding all the wide balls for calculating the number of balls faced, by including all the runs which are 4s and boundaries for calculating the number of 4s scored and similarly for 6s.

In [145]:
dfdotball = Ball_by_ball[(Ball_by_ball['Runs_Scored']!= 0)].groupby('Bowler')['MatcH_id'].count().reset_index()


In [153]:
dfdotball

Bowler  MatcH_id
0         1       195
1         4       225
2         5        35
3         8       200
4         9      1106
..      ...       ...
349     492        17
350     493        51
351     494        15
352     495        62
353     497        95

[354 rows x 2 columns]

In [154]:
dfdotball.columns=['Bowler','Non_dotball']


In [155]:
dfdotball

Bowler  Non_dotball
0         1          195
1         4          225
2         5           35
3         8          200
4         9         1106
..      ...          ...
349     492           17
350     493           51
351     494           15
352     495           62
353     497           95

[354 rows x 2 columns]

In [101]:
A={'a':"1,2,3,4"}
B=A['a'!=[2,3]]
print(B)


KeyError: True

In [129]:
dfBowling = Ball_by_ball.groupby('Bowler').agg({'MatcH_id':'nunique','Runs_Scored':'sum','Bowler_Wicket':'sum','Extra_runs':'sum','Ball_id':'count'}).reset_index()

In [130]:
dfBowling

Bowler  MatcH_id  Runs_Scored  Bowler_Wicket  Extra_runs  Ball_id
0         1        20          359             10          11      280
1         4        26          467              8          18      323
2         5         4           61              2          10       63
3         8        26          354              4          17      264
4         9        89         2222             65         126     1799
..      ...       ...          ...            ...         ...      ...
350     492         1           42              1           0       24
351     493         6          109              2           7       82
352     494         1           33              0           0       21
353     495         5          144              5          13      111
354     497        10          183              8           6      182

[355 rows x 6 columns]

In [132]:
dfBowling.columns =['Bowler','bowl_inning','Runs_conceded','bowler_wicket','bowler_extraruns','ball_throw']

In [133]:
dfBowling

Bowler  bowl_inning  Runs_conceded  bowler_wicket  bowler_extraruns  \
0         1           20            359             10                11   
1         4           26            467              8                18   
2         5            4             61              2                10   
3         8           26            354              4                17   
4         9           89           2222             65               126   
..      ...          ...            ...            ...               ...   
350     492            1             42              1                 0   
351     493            6            109              2                 7   
352     494            1             33              0                 0   
353     495            5            144              5                13   
354     497           10            183              8                 6   

     ball_throw  
0           280  
1           323  
2            63  
3           264  
4          1799  
..          ...  
350          24  
351          82  
352          21  
353         111  
354         182  

[355 rows x 6 columns]

## merging dfdotball and dfBowling and adding new column

In [157]:
dfBowling2=dfBowling.merge(dfdotball,on='Bowler',how='left')

In [158]:
dfBowling2

Bowler  bowl_inning  Runs_conceded  bowler_wicket  bowler_extraruns  \
0         1           20            359             10                11   
1         4           26            467              8                18   
2         5            4             61              2                10   
3         8           26            354              4                17   
4         9           89           2222             65               126   
..      ...          ...            ...            ...               ...   
350     492            1             42              1                 0   
351     493            6            109              2                 7   
352     494            1             33              0                 0   
353     495            5            144              5                13   
354     497           10            183              8                 6   

     ball_throw  Non_dotball  
0           280        195.0  
1           323        225.0  
2            63         35.0  
3           264        200.0  
4          1799       1106.0  
..          ...          ...  
350          24         17.0  
351          82         51.0  
352          21         15.0  
353         111         62.0  
354         182         95.0  

[355 rows x 7 columns]

In [172]:
dfBowling2['dot_ball %']=(((dfBowling2['ball_throw']-dfBowling2['Non_dotball'])/dfBowling2['ball_throw'])*100).round(decimals=2)

In [173]:
dfBowling2

Bowler  bowl_inning  Runs_conceded  bowler_wicket  bowler_extraruns  \
0         1           20            359             10                11   
1         4           26            467              8                18   
2         5            4             61              2                10   
3         8           26            354              4                17   
4         9           89           2222             65               126   
..      ...          ...            ...            ...               ...   
350     492            1             42              1                 0   
351     493            6            109              2                 7   
352     494            1             33              0                 0   
353     495            5            144              5                13   
354     497           10            183              8                 6   

     ball_throw  Non_dotball  dot_ball  dot_ball %  bowling_avg  
0           280        195.0      85.0       30.36        35.90  
1           323        225.0      98.0       30.34        58.38  
2            63         35.0      28.0       44.44        30.50  
3           264        200.0      64.0       24.24        88.50  
4          1799       1106.0     693.0       38.52        34.18  
..          ...          ...       ...         ...          ...  
350          24         17.0       7.0       29.17        42.00  
351          82         51.0      31.0       37.80        54.50  
352          21         15.0       6.0       28.57          inf  
353         111         62.0      49.0       44.14        28.80  
354         182         95.0      87.0       47.80        22.88  

[355 rows x 10 columns]

In [169]:
dfBowling2['bowling_avg']=(dfBowling2['Runs_conceded']/dfBowling2['bowler_wicket']).round(decimals=2)

In [174]:
dfBowling2

Bowler  bowl_inning  Runs_conceded  bowler_wicket  bowler_extraruns  \
0         1           20            359             10                11   
1         4           26            467              8                18   
2         5            4             61              2                10   
3         8           26            354              4                17   
4         9           89           2222             65               126   
..      ...          ...            ...            ...               ...   
350     492            1             42              1                 0   
351     493            6            109              2                 7   
352     494            1             33              0                 0   
353     495            5            144              5                13   
354     497           10            183              8                 6   

     ball_throw  Non_dotball  dot_ball  dot_ball %  bowling_avg  
0           280        195.0      85.0       30.36        35.90  
1           323        225.0      98.0       30.34        58.38  
2            63         35.0      28.0       44.44        30.50  
3           264        200.0      64.0       24.24        88.50  
4          1799       1106.0     693.0       38.52        34.18  
..          ...          ...       ...         ...          ...  
350          24         17.0       7.0       29.17        42.00  
351          82         51.0      31.0       37.80        54.50  
352          21         15.0       6.0       28.57          inf  
353         111         62.0      49.0       44.14        28.80  
354         182         95.0      87.0       47.80        22.88  

[355 rows x 10 columns]

In [176]:
dfBowling2['bowling_economy']=(dfBowling2['Runs_conceded']/(dfBowling2['ball_throw']/6)).round(decimals=2)

In [177]:
dfBowling2

Bowler  bowl_inning  Runs_conceded  bowler_wicket  bowler_extraruns  \
0         1           20            359             10                11   
1         4           26            467              8                18   
2         5            4             61              2                10   
3         8           26            354              4                17   
4         9           89           2222             65               126   
..      ...          ...            ...            ...               ...   
350     492            1             42              1                 0   
351     493            6            109              2                 7   
352     494            1             33              0                 0   
353     495            5            144              5                13   
354     497           10            183              8                 6   

     ball_throw  Non_dotball  dot_ball  dot_ball %  bowling_avg  \
0           280        195.0      85.0       30.36        35.90   
1           323        225.0      98.0       30.34        58.38   
2            63         35.0      28.0       44.44        30.50   
3           264        200.0      64.0       24.24        88.50   
4          1799       1106.0     693.0       38.52        34.18   
..          ...          ...       ...         ...          ...   
350          24         17.0       7.0       29.17        42.00   
351          82         51.0      31.0       37.80        54.50   
352          21         15.0       6.0       28.57          inf   
353         111         62.0      49.0       44.14        28.80   
354         182         95.0      87.0       47.80        22.88   

     bowling_economy  
0               7.69  
1               8.67  
2               5.81  
3               8.05  
4               7.41  
..               ...  
350            10.50  
351             7.98  
352             9.43  
353             7.78  
354             6.03  

[355 rows x 11 columns]

In [178]:
dfBowling2['bowling_strikerate']=(dfBowling2['ball_throw']/dfBowling2['bowler_wicket']).round(decimals=2)

In [226]:
dfBowling2

Bowler  bowl_inning  Runs_conceded  bowler_wicket  bowler_extraruns  \
0         1           20            359             10                11   
1         4           26            467              8                18   
2         5            4             61              2                10   
3         8           26            354              4                17   
4         9           89           2222             65               126   
..      ...          ...            ...            ...               ...   
350     492            1             42              1                 0   
351     493            6            109              2                 7   
352     494            1             33              0                 0   
353     495            5            144              5                13   
354     497           10            183              8                 6   

     ball_throw  Non_dotball  dot_ball  dot_ball %  bowling_avg  \
0           280        195.0      85.0       30.36        35.90   
1           323        225.0      98.0       30.34        58.38   
2            63         35.0      28.0       44.44        30.50   
3           264        200.0      64.0       24.24        88.50   
4          1799       1106.0     693.0       38.52        34.18   
..          ...          ...       ...         ...          ...   
350          24         17.0       7.0       29.17        42.00   
351          82         51.0      31.0       37.80        54.50   
352          21         15.0       6.0       28.57          inf   
353         111         62.0      49.0       44.14        28.80   
354         182         95.0      87.0       47.80        22.88   

     bowling_economy  bowling_strikerate  Batter  
0               7.69               28.00       1  
1               8.67               40.38       4  
2               5.81               31.50       5  
3               8.05               66.00       8  
4               7.41               27.68       9  
..               ...                 ...     ...  
350            10.50               24.00     492  
351             7.98               41.00     493  
352             9.43                 inf     494  
353             7.78               22.20     495  
354             6.03               22.75     497  

[355 rows x 13 columns]

In [225]:
dfBowling2['Batter']=dfBowling['Bowler']


In [180]:
dfBowling2.to_csv('player_bowling_dataset')

In [227]:
## now merge dfBowling2 and dfBall4(batting dataset
dfplayer=dfBall4.merge(dfBowling2,on='Batter',how='left')

In [228]:
dfplayer

Batter  Innings  Total_Runs  ball_faced  batting_strike_rate  out_inning  \
0         1       56        1349        1211               111.40        54.0   
1         2      103        2755        1971               139.78       100.0   
2         3        9          91         116                78.45         8.0   
3         4       61        1322        1039               127.24        49.0   
4         5        8          64          79                81.01         8.0   
..      ...      ...         ...         ...                  ...         ...   
455     490        1           3           2               150.00         NaN   
456     491        3          38          31               122.58         2.0   
457     495        3           8          10                80.00         3.0   
458     496        3          19          26                73.08         3.0   
459     497        5           9           9               100.00         NaN   

     batting_avg        player_name player_country  runs_from_fours  ...  \
0          24.98         SC Ganguly          India            548.0  ...   
1          27.55        BB McCullum    New Zealand           1108.0  ...   
2          11.38         RT Ponting      Australia             20.0  ...   
3          26.98          DJ Hussey      Australia            360.0  ...   
4           8.00    Mohammad Hafeez       Pakistan             28.0  ...   
..           ...                ...            ...              ...  ...   
455          NaN          RD Chahar          India              NaN  ...   
456        19.00         SP Jackson          India              NaN  ...   
457         2.67           TS Mills        England              NaN  ...   
458         6.33       Vishnu Vinod          India              4.0  ...   
459          NaN  Washington Sundar          India              NaN  ...   

     Runs_conceded  bowler_wicket  bowler_extraruns  ball_throw  Non_dotball  \
0            359.0           10.0              11.0       280.0        195.0   
1              NaN            NaN               NaN         NaN          NaN   
2              NaN            NaN               NaN         NaN          NaN   
3            467.0            8.0              18.0       323.0        225.0   
4             61.0            2.0              10.0        63.0         35.0   
..             ...            ...               ...         ...          ...   
455           54.0            2.0               5.0        46.0         25.0   
456            NaN            NaN               NaN         NaN          NaN   
457          144.0            5.0              13.0       111.0         62.0   
458            NaN            NaN               NaN         NaN          NaN   
459          183.0            8.0               6.0       182.0         95.0   

     dot_ball  dot_ball %  bowling_avg  bowling_economy  bowling_strikerate  
0        85.0       30.36        35.90             7.69               28.00  
1         NaN         NaN          NaN              NaN                 NaN  
2         NaN         NaN          NaN              NaN                 NaN  
3        98.0       30.34        58.38             8.67               40.38  
4        28.0       44.44        30.50             5.81               31.50  
..        ...         ...          ...              ...                 ...  
455      21.0       45.65        27.00             7.04               23.00  
456       NaN         NaN          NaN              NaN                 NaN  
457      49.0       44.14        28.80             7.78               22.20  
458       NaN         NaN          NaN              NaN                 NaN  
459      87.0       47.80        22.88             6.03               22.75  

[460 rows x 24 columns]

In [230]:
dfplayer2=dfBall4.merge(dfBowling2,on='Batter',how='outer')

In [231]:
dfplayer2

Batter  Innings  Total_Runs  ball_faced  batting_strike_rate  out_inning  \
0         1     56.0      1349.0      1211.0               111.40        54.0   
1         2    103.0      2755.0      1971.0               139.78       100.0   
2         3      9.0        91.0       116.0                78.45         8.0   
3         4     61.0      1322.0      1039.0               127.24        49.0   
4         5      8.0        64.0        79.0                81.01         8.0   
..      ...      ...         ...         ...                  ...         ...   
488     486      NaN         NaN         NaN                  NaN         NaN   
489     487      NaN         NaN         NaN                  NaN         NaN   
490     492      NaN         NaN         NaN                  NaN         NaN   
491     493      NaN         NaN         NaN                  NaN         NaN   
492     494      NaN         NaN         NaN                  NaN         NaN   

     batting_avg      player_name player_country  runs_from_fours  ...  \
0          24.98       SC Ganguly          India            548.0  ...   
1          27.55      BB McCullum    New Zealand           1108.0  ...   
2          11.38       RT Ponting      Australia             20.0  ...   
3          26.98        DJ Hussey      Australia            360.0  ...   
4           8.00  Mohammad Hafeez       Pakistan             28.0  ...   
..           ...              ...            ...              ...  ...   
488          NaN              NaN            NaN              NaN  ...   
489          NaN              NaN            NaN              NaN  ...   
490          NaN              NaN            NaN              NaN  ...   
491          NaN              NaN            NaN              NaN  ...   
492          NaN              NaN            NaN              NaN  ...   

     Runs_conceded  bowler_wicket  bowler_extraruns  ball_throw  Non_dotball  \
0            359.0           10.0              11.0       280.0        195.0   
1              NaN            NaN               NaN         NaN          NaN   
2              NaN            NaN               NaN         NaN          NaN   
3            467.0            8.0              18.0       323.0        225.0   
4             61.0            2.0              10.0        63.0         35.0   
..             ...            ...               ...         ...          ...   
488          209.0           10.0               5.0       141.0         88.0   
489           14.0            1.0               4.0        25.0          8.0   
490           42.0            1.0               0.0        24.0         17.0   
491          109.0            2.0               7.0        82.0         51.0   
492           33.0            0.0               0.0        21.0         15.0   

     dot_ball  dot_ball %  bowling_avg  bowling_economy  bowling_strikerate  
0        85.0       30.36        35.90             7.69               28.00  
1         NaN         NaN          NaN              NaN                 NaN  
2         NaN         NaN          NaN              NaN                 NaN  
3        98.0       30.34        58.38             8.67               40.38  
4        28.0       44.44        30.50             5.81               31.50  
..        ...         ...          ...              ...                 ...  
488      53.0       37.59        20.90             8.89               14.10  
489      17.0       68.00        14.00             3.36               25.00  
490       7.0       29.17        42.00            10.50               24.00  
491      31.0       37.80        54.50             7.98               41.00  
492       6.0       28.57          inf             9.43                 inf  

[493 rows x 24 columns]

In [229]:
dfplayer.to_csv('player_dataset')

In [233]:
dfplayer2.to_csv('player_dataset2')

### work on player_match dataset

In [234]:
Player_match

Player_match_SK  PlayerMatch_key  Match_Id  Player_Id  \
0                12694     3.359870e+10    335987          6   
1                12695     3.359870e+10    335987          7   
2                12696     3.359870e+10    335987          8   
3                12697     3.359870e+10    335987          9   
4                12698     3.359870e+10    335987         10   
...                ...              ...       ...        ...   
13987            26681     1.080000e+11   1082650        413   
13988            26682     1.080000e+11   1082650        430   
13989            26683     1.080000e+11   1082650        483   
13990            26684     1.080000e+11   1082650        488   
13991            26685     1.080000e+11   1082650        497   

             Player_Name         DOB     Batting_hand           Bowling_skill  \
0               R Dravid  01-11-1973   Right-hand bat      Right-arm offbreak   
1               W Jaffer   2/16/1978   Right-hand bat      Right-arm offbreak   
2                V Kohli  11-05-1988   Right-hand bat        Right-arm medium   
3              JH Kallis  10/16/1975   Right-hand bat   Right-arm fast-medium   
4               CL White   8/18/1983   Right-hand bat         Legbreak googly   
...                  ...         ...              ...                     ...   
13987          KH Pandya   3/24/1991    Left-hand bat  Slow left-arm orthodox   
13988            A Zampa   3/31/1992   Right-hand bat         Legbreak googly   
13989        LH Ferguson   6/13/1991   Right-hand bat          Right-arm fast   
13990        RA Tripathi  03-02-1991  ýRight-hand bat        Right-arm medium   
13991  Washington Sundar  10-05-1999   ýLeft-hand bat     ýRight-arm offbreak   

       Country_Name Role_Desc                  Player_team  \
0             India   Captain  Royal Challengers Bangalore   
1             India    Player  Royal Challengers Bangalore   
2             India    Player  Royal Challengers Bangalore   
3      South Africa    Player  Royal Challengers Bangalore   
4         Australia    Player  Royal Challengers Bangalore   
...             ...       ...                          ...   
13987         India    Player               Mumbai Indians   
13988     Australia    Player      Rising Pune Supergiants   
13989   New Zealand    Player      Rising Pune Supergiants   
13990         India    Player      Rising Pune Supergiants   
13991         India    Player      Rising Pune Supergiants   

                  Opposit_Team  Season_year  is_manofThematch  \
0        Kolkata Knight Riders         2008                 0   
1        Kolkata Knight Riders         2008                 0   
2        Kolkata Knight Riders         2008                 0   
3        Kolkata Knight Riders         2008                 0   
4        Kolkata Knight Riders         2008                 0   
...                        ...          ...               ...   
13987  Rising Pune Supergiants         2017                 1   
13988           Mumbai Indians         2017                 0   
13989           Mumbai Indians         2017                 0   
13990           Mumbai Indians         2017                 0   
13991           Mumbai Indians         2017                 0   

       Age_As_on_match  IsPlayers_Team_won Player_Captain Opposit_captain  \
0                   35                   0       R Dravid      SC Ganguly   
1                   30                   0       R Dravid      SC Ganguly   
2                   20                   0       R Dravid      SC Ganguly   
3                   33                   0       R Dravid      SC Ganguly   
4                   25                   0       R Dravid      SC Ganguly   
...                ...                 ...            ...             ...   
13987               26                   1      RG Sharma       SPD Smith   
13988               25                   0      SPD Smith       RG Sharma   
13989               26                   0    

In [243]:
dfmatch  = Player_match.groupby('Player_Id').agg({'Match_Id':'nunique','is_manofThematch':'sum','DOB':'unique'}).reset_index()

In [245]:
dfmatch

Batter  total_match_play  ismom           DOB
0         1                59      5  [07-08-1972]
1         2               103      5   [9/27/1981]
2         3                10      0  [12/19/1974]
3         4                64      2   [7/15/1977]
4         5                 8      0  [10/17/1980]
..      ...               ...    ...           ...
492     493                 6      0   [5/27/1991]
493     494                 1      0  [02-01-1996]
494     495                 5      0  [08-12-1992]
495     496                 3      0  [12-02-1993]
496     497                11      0  [10-05-1999]

[497 rows x 4 columns]

In [244]:
dfmatch.columns=['Batter','total_match_play','ismom','DOB']

In [246]:
dfplayer3=dfplayer2.merge(dfmatch,on='Batter',how='outer')


In [247]:
dfplayer3

Batter  Innings  Total_Runs  ball_faced  batting_strike_rate  out_inning  \
0         1     56.0      1349.0      1211.0               111.40        54.0   
1         2    103.0      2755.0      1971.0               139.78       100.0   
2         3      9.0        91.0       116.0                78.45         8.0   
3         4     61.0      1322.0      1039.0               127.24        49.0   
4         5      8.0        64.0        79.0                81.01         8.0   
..      ...      ...         ...         ...                  ...         ...   
492     494      NaN         NaN         NaN                  NaN         NaN   
493     466      NaN         NaN         NaN                  NaN         NaN   
494     467      NaN         NaN         NaN                  NaN         NaN   
495     468      NaN         NaN         NaN                  NaN         NaN   
496     469      NaN         NaN         NaN                  NaN         NaN   

     batting_avg      player_name player_country  runs_from_fours  ...  \
0          24.98       SC Ganguly          India            548.0  ...   
1          27.55      BB McCullum    New Zealand           1108.0  ...   
2          11.38       RT Ponting      Australia             20.0  ...   
3          26.98        DJ Hussey      Australia            360.0  ...   
4           8.00  Mohammad Hafeez       Pakistan             28.0  ...   
..           ...              ...            ...              ...  ...   
492          NaN              NaN            NaN              NaN  ...   
493          NaN              NaN            NaN              NaN  ...   
494          NaN              NaN            NaN              NaN  ...   
495          NaN              NaN            NaN              NaN  ...   
496          NaN              NaN            NaN              NaN  ...   

     ball_throw  Non_dotball  dot_ball  dot_ball %  bowling_avg  \
0         280.0        195.0      85.0       30.36        35.90   
1           NaN          NaN       NaN         NaN          NaN   
2           NaN          NaN       NaN         NaN          NaN   
3         323.0        225.0      98.0       30.34        58.38   
4          63.0         35.0      28.0       44.44        30.50   
..          ...          ...       ...         ...          ...   
492        21.0         15.0       6.0       28.57          inf   
493         NaN          NaN       NaN         NaN          NaN   
494         NaN          NaN       NaN         NaN          NaN   
495         NaN          NaN       NaN         NaN          NaN   
496         NaN          NaN       NaN         NaN          NaN   

     bowling_economy  bowling_strikerate  total_match_play  ismom  \
0               7.69               28.00                59      5   
1                NaN                 NaN               103      5   
2                NaN                 NaN                10      0   
3               8.67               40.38                64      2   
4               5.81               31.50                 8      0   
..               ...                 ...               ...    ...   
492             9.43                 inf                 1      0   
493              NaN                 NaN                 1      0   
494              NaN                 NaN                 1      0   
495              NaN                 NaN                 1      0   
496              NaN                 NaN                 1      0   

              DOB  
0    [07-08-1972]  
1     [9/27/1981]  
2    [12/19/1974]  
3     [7/15/1977]  
4    [10/17/1980]  
..            ...  
492  [02-01-1996]  
493   [9/27/1974]  
494  [09-06-1979]  
495  [12/14/1989]  
496  [12/22/1986]  

[497 rows x 27 columns]

In [248]:
dfplayer3.to_csv('player_dataset3')